### ISSUE TITLE
lein uberjar never finishes

### ISSUE THREAD

**Issue by vlaaad** [cljfx/cljfx#17]
Quoting @victorb:

@vlaaad another thing. Seems `lein uberjar` never finishes it's build, unsure of why. Seems there is maybe some issue with when clojure runs the classes static initializer and something gets started, which never gets closed afterwards.

Copying the example you made above:

```clojure
(ns lein-cljfx-test.core
  (:require [cljfx.api :as fx])
  (:gen-class)
  (:import [javafx.application Platform]))

(defn -main [& args]
  (Platform/setImplicitExit true)
  (fx/on-fx-thread
    (fx/create-component
      {:fx/type :stage
       :showing true
       :scene {:fx/type :scene
               :root {:fx/type :v-box
                      :children [{:fx/type :label
                                  :text "Hello from lein!"}]}}})))
```

Runs fine with `lein run` (and thanks to `setImplicitExit`, also exits the full application when closing the javafx application)

However, running `lein uberjar` with that example, leads to the lein getting stuck at `Compiling lein-cljfx-test.core` forever (seemingly, I let it run for 20 minutes)

Running `strace -f lein uberjar` also confirms that it's waiting for something, but that something never happens.

Any advice on what could be happening?

(using clojure 1.10.0 + cljfx 1.2.10 with both openjdk and jdk version 11)

**Comment by vlaaad** [comment]
@victorb this is because cljfx calls `(Platform/startup ...)`, which starts JavaFX application thread, and it's not a daemon. The problem is, without doing it compilation will fail. It's weird, but it works like this: to compile code clojure compiler has to load (import) JavaFX classes. Importing any class extending Control will throw an exception that application is not initialized, because in [Control](https://github.com/javafxports/openjdk-jfx/blob/develop/modules/javafx.controls/src/main/java/javafx/scene/control/Control.java)'s static initializer it does some styling-related settings:
```java
        // Ensures that the default application user agent stylesheet is loaded
        if (Application.getUserAgentStylesheet() == null) {
            PlatformImpl.setDefaultPlatformUserAgentStylesheet();
        }
```
And this requires JavaFX to be started.

I haven't found any solution to this, but a workaround is to have a different compile task that will explicitly call `(Platform/exit)` when it's done. It should be possible, since custom lein tasks are basically a clj files.

**Comment by vlaaad** [comment]
I'm currently busy with selection model stuff from #15, so I probably won't have time in the next couple of days to come up with example of such compile task, please share it if you will do it :+1:

**Comment by ahungry** [comment]
This worked as a fix for me, in project.clj add the following:

```
   :profiles {:uberjar
              {:aot :all
	       :injections [(javafx.application.Platform/exit)]}}
```

It seems to work fine still with `lein run` as the injection only happens during the uberjar step.
The gui compiles in a few seconds with no blocking for 20 minutes, and the uberjar itself runs perfectly.

**Comment by ahungry** [comment]
You could probably also use injections in this manner to tune the functionality of `(Platform/setImplicitExit true)` so that it does not exit on REPL usage, but does on real jar usage only.

**Comment by frenchy64** [comment]
Only starting JavaFX when `*compile-files*` is false seems to successfully compile the OP's example.

```clojure
(defn initialize []
  (if *compile-files*
    ::initialized
    (try
      (Platform/startup #(Platform/setImplicitExit false))
      ::initialized
      (catch IllegalStateException _
        ::already-initialized))))
```

**Comment by jbwyatt4** [comment]
Is there a solution for this issue? I am still encountering it on my recent attempts to make an uberjar.

**Comment by mhavrlent** [comment]
None of these workarounds work for me.
Workaround provided by "ahungry" ends up with this error:
`> lein uberjar
Compiling lein-cljfx-test.core
Syntax error macroexpanding at (api.clj:70:1).
Execution error (IllegalStateException) at com.sun.javafx.application.PlatformImpl/startup (PlatformImpl.java:145).
Platform.exit has been called
`
Full report at:
C:\Users\user\AppData\Local\Temp\clojure-5282345381059062488.edn
Compilation failed: Subprocess failed

Second workaround provided by "frenchy64" doesn't work. lein run keeps on running forever, no window pops up, lein uberjar keeps on running forever.

I'm using exactly the same code as "vlaaad" in the first post.
I also tried with removed "(Platform/setImplicitExit true)" line, same result.

I'm going to try deps.edn now, hopefully that will work....

**Comment by vlaaad** [comment]
Hi @mhavrlent 
Can you share a repro lein project so I can have a look? Also, what version of Java are you using?

**Comment by mhavrlent** [comment]
@vlaaad 
Here's the repo: https://github.com/mhavrlent/lein-cljfx-test
I tried with both Java 1.8.0_231 and 11.0.3_7, same results.

**Comment by vlaaad** [comment]
Thanks a lot, I'll have a look!

**Comment by vlaaad** [comment]
Should be solved with 1.7.0, see [AOT](https://github.com/cljfx/cljfx#aot-compilation-is-complicated) readme section.

**Comment by jdf-id-au** [comment]
I had `:injections [(javafx.application.Platform/exit)]` working in my leiningen uberjar profile. But I'm now using a `(proxy [DateCell] ...` (with cljfx 1.7.4) and get _Unexpected error (IllegalStateException) macroexpanding proxy_ because _Toolkit not initialized_. Is there a way of getting Leiningen to run the injection *after* all the compiling's done?
